### ISSUE TITLE
relative target path for intersphinx leads to broken links

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1204]
The intersphinx docs say: "Relative local paths for target locations are taken as relative to the base of the built documentation". But that doesn't work when the document is in a subdirectory. The href should have `../` prepended as many times as needed, but it doesn't.

For example, an intersphinx mapping like this:

``` python
intersphinx_mapping = {
    'other': ('../other/', '../other/other.inv'),
}
```

when used in a document such as topic/details.rst, will link to `../other/something` whereas it should link to `../../other/something`.

I had a go at fixing this myself, I got as far as figuring out that it should be calling `app.builder.get_relative_uri`, something like this:

``` diff
--- a/sphinx/ext/intersphinx.py
+++ b/sphinx/ext/intersphinx.py
@@ -220,7 +220,8 @@ def missing_reference(app, env, node, contnode):
             if objtype not in inventory or target not in inventory[objtype]:
                 continue
             proj, version, uri, dispname = inventory[objtype][target]
-            newnode = nodes.reference('', '', internal=False, refuri=uri,
+            newnode = nodes.reference('', '', internal=False,
+                          refuri=app.builder.get_relative_uri('WHAT GOES HERE?', uri),
                           reftitle=_('(in &#37;s v&#37;s)') &#37; (proj, version))
             if node.get('refexplicit'):
                 # use whatever title was given
```

but I can't figure out how the missing_reference hook can tell what the source document path is (I was expecting I could do something like node.document.name but that is not a thing).

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1204
- Originally reported by: Dan Callaghan
- Originally created at: 2013-06-26T08:31:31.376

**Comment by shimizukawa** [comment]
_From WAKAYAMA shirou on 2013-09-16 07:52:57+00:00_

Currently intersphinx assumes the base directory is the build output dir.

So I think you rewrite the intersphinx_mapping in the conf.py to relative path from the build output dir (../../other/something).

If add get_relative_uri() by your suggestion, some other documents may become broken.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-09-16 08:22:04+00:00_

The issue is valid: intersphinx links from a.rst and b/c.rst end up having different base paths.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2013-09-16 08:34:09+00:00_

Fixes #1204: use the correct relative path referring to local intersphinx targets.

→ <<cset 0551a8731c2b33eb3b28a72177f6eb79efd6a1bb>>
